# Анализ базы данных (SQL)

**Цель** - проанализировать базу данных купленного сервиса для чтения книг по подписке для дальнейшего формулирования ценностного предложения для нового продукта.

**Задачи:**

* познакомиться с хранящимися в базе данными,
* проанализировать результаты запросов.

Данные расположены в 5 таблицах, содержащих информацию о книгах, издательствах, авторах, пользовательских оценках и обзорах книг.

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


## Изучение данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
pd.set_option('display.float_format', '{:.2f}'.format)
pd.set_option('display.max_columns', None)

In [3]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [4]:
#pd.io.sql.read_sql(query, con = engine) 

In [5]:
def first_look_sql(table):    
    display(pd.io.sql.read_sql('''SELECT * FROM {} LIMIT 5'''.format(table), con = engine))
    display()
    display(pd.io.sql.read_sql('''SELECT * FROM {}'''.format(table), con = engine).info())
    display()
    display(pd.io.sql.read_sql('''SELECT * FROM {}'''.format(table), con = engine).describe())

In [6]:
first_look_sql('books')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

book_id  author_id  num_pages  publisher_id
count  1000.00    1000.00    1000.00       1000.00
mean    500.50     320.42     389.11        171.27
std     288.82     181.62     229.39         99.08
min       1.00       1.00      14.00          1.00
25%     250.75     162.75     249.00         83.00
50%     500.50     316.50     352.00        177.50
75%     750.25     481.00     453.00        258.00
max    1000.00     636.00    2690.00        340.00

In [7]:
first_look_sql('authors')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

author_id
count     636.00
mean      318.50
std       183.74
min         1.00
25%       159.75
50%       318.50
75%       477.25
max       636.00

In [8]:
first_look_sql('publishers')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

publisher_id
count        340.00
mean         170.50
std           98.29
min            1.00
25%           85.75
50%          170.50
75%          255.25
max          340.00

In [9]:
first_look_sql('ratings')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

rating_id  book_id  rating
count    6456.00  6456.00 6456.00
mean     3228.50   510.57    3.93
std      1863.83   284.14    0.94
min         1.00     1.00    1.00
25%      1614.75   291.00    3.00
50%      3228.50   506.00    4.00
75%      4842.25   750.00    5.00
max      6456.00  1000.00    5.00

In [10]:
first_look_sql('reviews')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

review_id  book_id
count    2793.00  2793.00
mean     1397.00   504.69
std       806.41   288.47
min         1.00     1.00
25%       699.00   259.00
50%      1397.00   505.00
75%      2095.00   753.00
max      2793.00  1000.00

Подключились к базе данных PostgreSQL, ознакомились с содержимым 5 таблиц: 
* books, 
* authors, 
* publishers, 
* ratings, 
* reviews.

## Задания

### Кол-во книг после 1 января 2000 года

In [11]:
query = ''' 
            SELECT COUNT(book_id) AS books_count
            FROM books
            WHERE publication_date >= DATE('2000-01-01')
        '''

In [12]:
pd.io.sql.read_sql(query, con = engine)

books_count
0          821

### Кол-во обзоров и средняя оценка каждой книги


In [13]:
query = ''' 
            SELECT title, 
                   SUM(rev_sum) as rev_sum,
                   AVG(avg_rate) as avg_rate  
            FROM books
            LEFT JOIN (SELECT book_id, 
                              COUNT(review_id) AS rev_sum
                       FROM reviews
                       GROUP BY book_id) AS one
            ON books.book_id = one.book_id
            LEFT JOIN (SELECT book_id, 
                              AVG(rating) AS avg_rate
                       FROM ratings
                       GROUP BY book_id) AS two
            ON books.book_id = two.book_id
            GROUP BY books.title
            
        '''

In [14]:
pd.io.sql.read_sql(query, con = engine)

title  rev_sum  avg_rate
0                            The Count of Monte Cristo     5.00      4.22
1                              Count Zero (Sprawl  #2)     2.00      2.50
2    The Botany of Desire: A Plant's-Eye View of th...     2.00      3.50
3                                 The Poisonwood Bible     5.00      4.36
4                                 The Canterbury Tales     3.00      3.33
..                                                 ...      ...       ...
994  In the Heart of the Sea: The Tragedy of the Wh...     3.00      3.33
995                           Of Love and Other Demons     2.00      4.50
996               Welcome to Temptation (Dempseys  #1)     2.00      5.00
997                      World's End (The Sandman  #8)     2.00      4.50
998                                  Holes (Holes  #1)     5.00      3.97

[999 rows x 3 columns]

### Издательство выпустившее наибольшее число книг толще 50 страниц

Таким образом исключим из анализа брошюры

In [15]:
query = ''' 
            SELECT publisher, 
                   COUNT(book_id) as books_count 
            FROM books 
            LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
            WHERE num_pages > 50
            GROUP BY publisher
            ORDER BY books_count DESC
            LIMIT 1            
        '''

In [16]:
pd.io.sql.read_sql(query, con = engine)

publisher  books_count
0  Penguin Books           42

### Fвтора с самой высокой средней оценкой книг (только книги с 50 и более оценками)

In [17]:
query = '''             
        SELECT author,
               ROUND(AVG(avg_raiting), 2) AS avg_raiting, 
        SUM(rating_count) AS rate_count
        FROM    
        (SELECT author_id, 
                title, 
                AVG(rating) AS avg_raiting, 
                COUNT(rating) AS rating_count
        FROM books LEFT JOIN ratings ON books.book_id = ratings.book_id 
        GROUP BY title, author_id
        HAVING COUNT(rating_id) >= 50) as one
        LEFT JOIN authors ON one.author_id = authors.author_id
        GROUP BY author
        ORDER BY avg_raiting DESC
        LIMIT 1                
        '''

In [18]:
pd.io.sql.read_sql(query, con = engine)

author  avg_raiting  rate_count
0  J.K. Rowling/Mary GrandPré         4.28      310.00

### Cреднее кол-во обзоров от пользователей, которые поставили больше 50 оценок

In [19]:
query = '''             
        SELECT ROUND(AVG(rev_counts)) AS avg_rev
        FROM
            (SELECT COUNT(review_id) AS rev_counts
            FROM
                (SELECT username 
                FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) > 50) AS sub
            LEFT JOIN reviews ON sub.username = reviews.username
            GROUP BY sub.username) AS sub_2
        '''

In [20]:
pd.io.sql.read_sql(query, con = engine)

avg_rev
0    24.00

## Общий вывод

**Анализ данных показал, что:**

* Количество книг, выпущенных после 1 января 2000 года, составляет **821** штуку.

* Посчитали **количество обзоров - 999**, среднюю оценку для каждой из книг 

* Издательство, которое выпустило **наибольшее число книг толще 50 страниц**, это **Penguin Books с 42** книгами (исключая брошюры)

* Автор с самой высокой средней оценкой книг - это **J.K. Rowling/Mary GrandPré** со средним рейтингом **4.28**, составленным по 310 оценкам пользователей

* В среднем **количество обзоров** от пользователей, которые **поставили больше 50 оценок**, составляет **24 штуки**
